File: 2025/python/day02.py
from utils import get_data
from utils import get_nums
from utils import timeit


def process_data(data):
    nums = get_nums(data, signed=False)
    return [(nums[i], nums[i+1]) for i in range(0, len(nums), 2)]


def p1_invalid(num):
    num_str = str(num)

    if len(num_str) % 2 != 0:
        return False

    return num_str[:len(num_str)//2] == num_str[len(num_str)//2:]


def p2_invalid(num):
    num_str = str(num)

    for prefix_len in range(1, len(num_str) // 2 + 1):
        prefix = num_str[:prefix_len]

        if len(num_str) % prefix_len == 0:
            repetitions = len(num_str) // prefix_len

            if repetitions >= 2 and prefix * repetitions == num_str:
                return True

    return False


@timeit
def main():
    data = get_data("02")
    data = process_data(data)

    p1 = 0
    p2 = 0
    for low, high in data:
        for num in range(low, high + 1):
            if p1_invalid(num):
                p1 += num
            if p2_invalid(num):
                p2 += num

    print(p1)
    print(p2)

File: 2025/rust/day02.rs
use crate::y2025::utils;

fn process_data(data: &str) -> Vec<(u64, u64)> {
    let nums = utils::get_nums_unsigned(data);
    nums.chunks_exact(2)
        .map(|chunk| (chunk[0], chunk[1]))
        .collect()
}

fn p1_invalid(num: u64) -> bool {
    let num_str = num.to_string();
    let len = num_str.len();

    if len % 2 != 0 {
        return false;
    }

    let half = len / 2;
    &num_str[..half] == &num_str[half..]
}

fn p2_invalid(num: u64) -> bool {
    let num_str = num.to_string();
    let len = num_str.len();

    for prefix_len in 1..=len / 2 {
        let prefix = &num_str[..prefix_len];

        if len % prefix_len == 0 {
            let repetitions = len / prefix_len;

            if repetitions >= 2 {
                let repeated = prefix.repeat(repetitions);
                if repeated == num_str {
                    return true;
                }
            }
        }
    }

    false
}

pub fn main() {
    let data = utils::get_data("02");
    let data = process_data(&data);

    let mut p1 = 0u64;
    let mut p2 = 0u64;

    for (low, high) in data {
        for num in low..=high {
            if p1_invalid(num) {
                p1 += num;
            }
            if p2_invalid(num) {
                p2 += num;
            }
        }
    }

    println!("{}", p1);
    println!("{}", p2);
}
